server/src/authority/message_response.rs

```diff
@@ -1,115 +1,165 @@
 
 /// A EncodableMessage with borrowed data for Responses in the Server
 #[derive(Debug)]
-pub struct MessageResponse<'q, 'a> {
+pub struct MessageResponse<
+    'q,
+    'a,
+    A = AuthLookup<'a, 'q>,
+    N = Chain<LookupRecords<'q, 'a>, LookupRecords<'q, 'a>>,
+> where
+    A: 'q + 'a + Iterator<Item = &'a Record>,
+    N: 'q + 'a + Iterator<Item = &'a Record>,
+{
     header: Header,
     queries: Option<&'q Queries<'q>>,
-    answers: Vec<&'a Record>,
-    name_servers: Vec<&'a Record>,
+    answers: A,
+    name_servers: N,
     additionals: Vec<&'a Record>,
     sig0: Vec<Record>,
     edns: Option<Edns>,
 }
 
-impl<'q, 'a> MessageResponse<'q, 'a> {
-    /// Constructs a new Response
-    ///
-    /// # Arguments
-    ///
-    /// * `queries` - any optional set of Queries to associate with the Response
-    pub fn new(queries: Option<&'q Queries<'q>>) -> MessageResponseBuilder<'q, 'a> {
-        MessageResponseBuilder {
-            queries,
-            answers: None,
-            name_servers: None,
-            additionals: None,
-            sig0: None,
-            edns: None,
-        }
+impl<'q, 'a, A, N> MessageResponse<'q, 'a, A, N>
+where
+    A: 'q + 'a + Iterator<Item = &'a Record>,
+    N: 'q + 'a + Iterator<Item = &'a Record>,
+{
+    /// Returns the header of the message
+    pub fn header(&self) -> &Header {
+        &self.header
     }
 
     /// Set the EDNS options for the Response
     pub fn set_edns(&mut self, edns: Edns) -> &mut Self {
         self.edns = Some(edns);
         self
     }
-}
-
-macro_rules! section {
-    ($s:ident, $l:ident, $e:ident) => {
-        fn $l(&self) -> usize {
-            self.$s.len()
-        }
-
-        fn $e(&self, encoder: &mut BinEncoder) -> ProtoResult<()> {
-            encoder.emit_all_refs(self.$s.iter())
-        }
-    }
-}
 
-impl<'q, 'a> EncodableMessage for MessageResponse<'q, 'a> {
-    fn header(&self) -> &Header {
-        &self.header
-    }
-
-    fn queries_len(&self) -> usize {
-        self.queries.map(|s| s.len()).unwrap_or(0)
-    }
-
-    fn emit_queries(&self, encoder: &mut BinEncoder) -> ProtoResult<()> {
+    fn emit_queries(&self, encoder: &mut BinEncoder) -> ProtoResult<usize> {
         if let Some(queries) = self.queries {
-            encoder.emit_vec(queries.as_bytes())
+            encoder
+                .emit_vec(queries.as_bytes())
+                .map(|()| self.queries.map(|s| s.len()).unwrap_or(0))
         } else {
-            Ok(())
+            Ok(0)
         }
     }
 
-    section!(answers, answers_len, emit_answers);
-    section!(name_servers, name_servers_len, emit_name_servers);
-    section!(additionals, additionals_len, emit_additionals);
-
     fn edns(&self) -> Option<&Edns> {
         self.edns.as_ref()
     }
 
     fn sig0(&self) -> &[Record] {
         &self.sig0
     }
+
+    /// Consumes self, and emits to the encoder.
+    pub fn destructive_emit(mut self, encoder: &mut BinEncoder) -> ProtoResult<()> {
```

Comment:
before merging this, I want to take another pass and see if this can be deduped with the `Message::emit` method.

Reply:
I found a solution to this. I'm just polishing up that change, and then we will have less code duplication... and less code (I think).